akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,147 @@
+      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
+      private[this] var aggregator: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        if (pullOnStart) pull(in)
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay("FoldWithinIntervalTimer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit =
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregator != null && isReadyOnTimer(aggregator)) harvestAndEmit()
+        }
+
+      override def onPush(): Unit = {
+        aggregateAndEmitIfReady()
+        if (isAvailable(out)) pull(in) // pull only if there is no backpressure from outlet
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        harvestAndEmit()
+        completeStage()
+      }
+
+      override def onPull(): Unit = {
+        aggregateAndEmitIfReady()
+        if (!hasBeenPulled(in)) pull(in) // always pass through the pull to upstream
+      }
+
+      setHandlers(in, out, this)
+
+      private def harvestAndEmit(): Unit =
+        if (aggregator != null) {
+          emit(out, FoldWith.this.harvest(aggregator))
+          aggregator = null.asInstanceOf[Agg]
+        }
+
+      private def aggregateAndEmitIfReady(): Unit = if (isAvailable(in)) {
+        val input = grab(in)
+        aggregator = if (aggregator == null) seed(input) else aggregate(aggregator, input)
+        if (emitOnAgg(aggregator)) harvestAndEmit()
+      }
+
+    }
+
+}
+
+/**
+ * This is a convenient wrapper of [[FoldWith]] to handle timing constraints
+ * @param seed        initiate the aggregated output with first input
+ * @param aggregate   sequentially aggregate input
+ * @param emitOnAgg   decide whether the current aggregator can be emitted, invoked after each aggregate
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage
+ *                    there can be extra time between harvest and next stage receiving the emitted output
+ *                    time sensitive operations can be added here, such as closing an output channel
+ * @param maxGap      the gap allowed between consecutive aggregate operations
+ * @param maxDuration the duration of the sequence of aggregate operations from initial seed until emit is triggered
+ * @param interval    interval of the timer to check the maxGap and maxDuration condition
+ * @param getSystemTimeMs source of the system time, in case of testing simulated time can be used
+ */
+class FoldWithin[In, Agg, Out](
+    seed: In => Agg,
+    aggregate: (Agg, In) => Agg,
+    emitOnAgg: Agg => Boolean,
+    harvest: Agg => Out,
+    maxGap: Option[FiniteDuration] = None,
+    maxDuration: Option[FiniteDuration] = None,
+    interval: FiniteDuration = 1.milli,
+    getSystemTimeMs: => Long = System.currentTimeMillis(),
```

Comment:
The timer only need ms level resolution in practice, and nanoTime is more expensive especially it's called for every data which could be a huge burden.
The good thing is this is a parameter now, we can add comments to let advanced users to make the decision.
We do need to pick a default time getter in the helper methods, which I feel this cheaper but less accurate one should be the default.

Reply:
It is not about accuracy but about correctness, using `System.currentTimeMillis` any case where you expect to measure time passed means you will not always measure time passed as it may move backwards. A user could make that tradeoff but we should not provide a default using it.

Reply:
removed all default parameter values